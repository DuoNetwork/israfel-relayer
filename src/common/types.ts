import { SignedOrder } from '0x.js';

export interface ILiveOrder {
	account: string;
	pair: string;
	orderHash: string;
	price: number;
	amount: number;
	balance: number;
	fill: number;
	side: string;
	createdAt?: number;
	updatedAt?: number;
	initialSequence: number;
	currentSequence: number;
}

export interface IUserOrder extends ILiveOrder {
	type: string;
	status: string;
	updatedBy: string;
}

export interface IRawOrder {
	orderHash: string;
	signedOrder: IStringSignedOrder | SignedOrder;
	createdAt?: number;
	updatedAt?: number;
}

export interface IStringSignedOrder {
	senderAddress: string;
	makerAddress: string;
	takerAddress: string;
	makerFee: string;
	takerFee: string;
	makerAssetAmount: string;
	takerAssetAmount: string;
	makerAssetData: string;
	takerAssetData: string;
	salt: string;
	exchangeAddress: string;
	feeRecipientAddress: string;
	expirationTimeSeconds: string;
	signature: string;
}

export interface IOrderPersistRequest {
	method: string;
	pair: string;
	orderHash: string;
	balance: number;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderQueueItem {
	method: string;
	liveOrder: ILiveOrder;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderBookSnapshotWs extends IOrderBookSnapshot {
	type: string;
	sequence: number;
	channel: string;
}

export interface IOrderBookSnapshot {
	sequence: number;
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IUpdatePayloadWs {
	order: IStringSignedOrder;
	metaData: {
		remainingTakerAssetAmount: string;
	};
}

export interface IOrderBookUpdateWS {
	price: number;
	amount: number;
}

export interface IOrderBookUpdate {
	sequence: number;
	pair: string;
	price: number;
	amount: number;
}

export interface IWsRequest {
	method: string;
	channel: string;
	pair: string;
}

export interface IWsResponse {
	status: string;
	method: string;
	channel: string;
	pair: string;
}

export interface IWsOrderRequest extends IWsRequest {
	orderHash: string;
}

export interface IWsAddOrderRequest extends IWsOrderRequest {
	order: IStringSignedOrder | SignedOrder;
}

export interface IWsOrderResponse extends IWsResponse {
	orderHash: string;
}

export interface IWsUserOrderResponse extends IWsOrderResponse {
	userOrder: IUserOrder;
}

export interface ISubscribeOrderBookRequest {
	method: string;
	channel: string;
}

// export interface IUpdateResponseWs {
// 	type: WsChannelResposnseTypes;
// 	lastTimestamp: number;
// 	currentTimestamp: number;
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	bids: IOrderBookUpdateWS[];
// 	asks: IOrderBookUpdateWS[];
// }

// export interface IOrderResponseWs {
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	status: string;
// 	failedReason: string;
// }

export interface IOption {
	live: boolean;
	token: string;
	maker: number;
	spender: number;
	amount: number;
	debug: boolean;
	server: boolean;
}

export interface IStatus {
	hostname: string;
	updatedAt: number;
	pair: string;
	tool: string;
	count?: number;
}

export interface IService {
	service: string;
	hostname: string;
	url: string;
}
