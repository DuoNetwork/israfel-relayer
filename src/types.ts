import { BigNumber, ExchangeContractErrs, OrderRelevantState, SignedOrder } from '0x.js';

export interface IDuoOrder extends IDuoSignedOrder {
	orderHash: string;
	isValid: boolean;
	isCancelled: boolean;
	updatedAt: number;
	orderWatcherState: OrderRelevantState | ExchangeContractErrs;
}

// export interface IOrderBook {
// 	bids: IDuoOrder[];
// 	asks: IDuoOrder[];
// }

export interface IOrderBookSnapshotWs {
	type: WsChannelResposnseTypes;
	timestamp: number;
	channel: {
		name: string;
		marketId: string;
	};
	requestId: number;
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IUpdatePayloadWs {
	order: SignedOrder;
	metaData: {
		remainingTakerAssetAmount: BigNumber;
	};
}

export interface IOrderBookUpdateWS {
	price: string;
	amount: string;
}

export interface IUpdateResponseWs {
	type: WsChannelResposnseTypes;
	lastTimestamp: number;
	currentTimestamp: number;
	channel: {
		name: WsChannelName;
		marketId: string;
	};
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IOrderResponseWs {
	channel: {
		name: WsChannelName;
		marketId: string;
	};
	status: string;
	failedReason: string;
}

export interface ICancelOrderResponseWs {
	status: string;
	orderHash: string;
}

export enum ErrorResponseWs {
	InvalidOrder = 'Invalid order schema or signature!',
	NoExistOrder = 'Order does not exist in DB!'
}

export interface IDuoSignedOrder {
	senderAddress: string;
	makerAddress: string;
	takerAddress: string;
	makerFee: string;
	takerFee: string;
	makerAssetAmount: string;
	takerAssetAmount: string;
	makerAssetData: string;
	takerAssetData: string;
	salt: string;
	exchangeAddress: string;
	feeRecipientAddress: string;
	expirationTimeSeconds: string;
	signature: string;
}

export enum WsChannelMessageTypes {
	Add = 'add',
	Cancel = 'cancel',
	Subscribe = 'subscribe'
}

export enum WsChannelResposnseTypes {
	Update = 'update',
	Snapshot = 'snapshot'
}

export enum WsChannelName {
	Orderbook = 'orderbook',
	Order = 'order'
}

export interface IOrderStateCancelled {
	isCancelled: boolean;
	orderHash: string;
}

export interface IOption {
	token: string;
	maker: number;
	spender: number;
	amount: number;
}
