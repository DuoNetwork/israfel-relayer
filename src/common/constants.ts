export const DB_ISRAFEL = 'israfel';
export const DB_LIVE_ORDERS = 'liveOrders';
export const DB_RAW_ORDERS = 'rawOrders';
export const DB_USER_ORDERS = 'userOrders';
export const DB_SEQUENCE = 'sequence';
export const DB_STATUS = 'status';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PAIR = 'pair';
export const DB_INITIAL_SEQ = 'initialSequence';
export const DB_CURRENT_SEQ = 'currentSequence';
export const DB_ORDER_HASH = 'orderHash';
export const DB_PRICE = 'price';
export const DB_ACCOUNT = 'account';
export const DB_ACCOUNT_YM = 'accountYearMonth';
export const DB_PAIR_OH_SEQ_STATUS = 'pairOrderHashSequenceStatus';
export const DB_TYPE = 'type';
export const DB_CONFIRMED = 'confirmed';
export const DB_MATCHING = 'matching';
export const DB_USER = 'user';
export const DB_ORDER_PROCESSOR = 'orderProcessor';
export const DB_ORDER_MATCHER = 'orderMatcher';
export const DB_ORDER_WATCHER = 'orderWatcher';
export const DB_RELAYER = 'relayer';
export const DB_URL = 'url';
export const DB_SIDE = 'side';
export const DB_ASK = 'ask';
export const DB_BID = 'bid';
export const DB_BALANCE = 'balance';
export const DB_AMOUNT = 'amount';
export const DB_FILL = 'fill';
export const DB_PFILL = 'pFill';
export const DB_CREATED_AT = 'createdAt';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_UPDATED_BY = 'updatedBy';
export const DB_PROCESSED = 'processed';
export const DB_TOKENS = 'tokens';
export const DB_ADDRESS = 'address';
export const DB_CODE = 'code';
export const DB_EXP = 'expiry';
export const DB_DENOMINATION = 'denomination';
export const DB_PRECISION = 'precision';
export const DB_FEE = 'fee';
export const DB_FEE_ASSET = 'feeAsset';

export const DB_ADD = 'add';
export const DB_UPDATE = 'update';
export const DB_TERMINATE = 'terminate';

export const DB_PROCESS = 'process';
export const DB_HOSTNAME = 'hostname';
export const DB_COUNT = 'count';

export const DB_0X_SENDER_ADDR = 'senderAddress';
export const DB_0X_MAKER_FEE = 'makerFee';
export const DB_0X_TAKER_FEE = 'takerFee';
export const DB_0X_MAKER_ADDR = 'makerAddress';
export const DB_0X_TAKER_ADDR = 'takerAddress';
export const DB_0X_MAKER_ASSET_AMT = 'makerAssetAmount';
export const DB_0X_TAKER_ASSET_AMT = 'takerAssetAmount';
export const DB_0X_MAKER_ASSET_DATA = 'makerAssetData';
export const DB_0X_TAKER_ASSET_DATA = 'takerAssetData';
export const DB_0X_SALT = 'salt';
export const DB_0X_EXCHANGE_ADDR = 'exchangeAddress';
export const DB_0X_FEE_RECIPIENT_ADDR = 'feeRecipientAddress';
export const DB_0X_EXPIRATION_TIME_SECONDS = 'expirationTimeSeconds';
export const DB_0X_SIGNATURE = 'signature';

export const DB_ORDERS = 'orders';
export const DB_ORDER_BOOKS = 'orderBooks';
export const DB_QUEUE = 'queue';
export const DB_CACHE = 'cache';
export const DB_PUBSUB = 'pubsub';
export const DB_SNAPSHOT = 'snapshot';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';

export const TAKER_ETH_DEPOSIT = 10; // for development only
export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const RELAYER_ADDR_KOVAN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const RELAYER_ADDR_MAIN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const DUMMY_ADDR = '0x0000000000000000000000000000000000000000';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const NETWORK_ID_MAIN = 1;
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const PENDING_HOURS = 24;

export const WS_HISTORY = 'history';
export const WS_INFO = 'info';
export const WS_SUB = 'subscribe';
export const WS_UNSUB = 'unsubscribe';
export const WS_CHANNEL = 'channel';
export const WS_METHOD = 'method';
export const WS_INVALID_REQ = 'invalid request';
export const WS_INVALID_ORDER = 'invalid order';
export const WS_ERROR = 'error';
export const WS_OK = 'ok';
export const WS_SERVICE_NA = 'service not available';

export const WS_CHANNEL_ORDERBOOK = 'orderbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const TOKEN_WETH = 'WETH';

export const PRUNE_INTERVAL = 3600000;
export const ORDER_PRUNE = 'pruneOrder';
export const SET_ALLOWANCE = 'setAllowance';
export const START_RELAYER = 'startRelayer';

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const PRICE_PRECISION = 8;

export const BASE_DERIVATION_PATH = `44'/60'/0'/0`;
