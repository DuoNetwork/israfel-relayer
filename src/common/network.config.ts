
export const mnemonicConfig = {
	MNEMONIC: 'alice bob charles david eric frank george henry is jack kobe let',
	BASE_DERIVATION_PATH: `44'/60'/0'/0`
}

export const KOVAN_CONFIGS = {
	rpcUrl: 'https://kovan.infura.io/WSDscoNUvMiL1M7TvMNP',
	networkId: 42,
};
