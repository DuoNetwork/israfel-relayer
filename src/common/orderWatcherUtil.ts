import { ContractWrappers, OrderWatcher, RPCSubprovider } from '0x.js';
import * as CST from '../constants';
import dynamoUtil from '../dynamoUtil';
import { providerEngine } from '../providerEngine';
import { ILiveOrders, IOption } from '../types';
import util from '../util';

class OrderWatcherUtil {
	public provider = new RPCSubprovider(CST.PROVIDER_LOCAL);
	// public providerEngine = new Web3ProviderEngine();
	public zeroEx: ContractWrappers;
	public orderWatcher: OrderWatcher;

	constructor() {
		// this.providerEngine.addProvider(this.provider);
		// this.providerEngine.start();
		this.zeroEx = new ContractWrappers(providerEngine, { networkId: CST.NETWORK_ID_LOCAL });
		this.orderWatcher = new OrderWatcher(providerEngine, CST.NETWORK_ID_LOCAL);
	}

	public init(tool: string, option: IOption) {
		const config = require('./keys/' + (option.live ? 'live' : 'dev') + '/dynamo.json');
		dynamoUtil.init(config, option.live, tool);
	}

	public unsubOrderWatcher() {
		this.orderWatcher.unsubscribe();
	}

	//remove orders remaining invalid for 24 hours from DB
	public async pruneOrders(option: IOption) {
		const marketId = option.token + '-' + CST.TOKEN_WETH;
		const orders = await dynamoUtil.getLiveOrders(marketId);
		console.log('length before prune is', orders.length);
		orders.forEach(order => {
			const inValidTime = !order.isValid ? Date.now() - order.updatedAt : 0;
			console.log(inValidTime);
			if (inValidTime > CST.PENDING_HOURS * 3600000) {
				dynamoUtil.removeLiveOrder(marketId, order.orderHash);
				this.orderWatcher.removeOrder(order.orderHash);
				console.log('remove order!');
			}
		});
		console.log('length after prune is', orders.length);
	}

	public async startOrderWatcher(option: IOption) {
		const marketId = option.token + '-' + CST.TOKEN_WETH;
		util.logInfo('start order watcher for ' + marketId);
		const liveOrders: ILiveOrders[] = await dynamoUtil.getLiveOrders(marketId);

		console.log('length in DB is ', liveOrders.length);
		for (const order of liveOrders)
			try {
				await this.orderWatcher.addOrderAsync(
					await dynamoUtil.getRawOrder(order.orderHash)
				);
				console.log('succsfully added %s', order.orderHash);
			} catch (e) {
				console.log('failed to add %s', order.orderHash, 'error is ' + e);
			}

		this.orderWatcher.subscribe(async (err, orderState) => {
			if (err) {
				console.log(err);
				return;
			}

			console.log(Date.now().toString(), orderState);
			if (orderState !== undefined) await dynamoUtil.updateOrderState(orderState, marketId);
		});
	}
}

const orderWatcherUtil = new OrderWatcherUtil();
export default orderWatcherUtil;
