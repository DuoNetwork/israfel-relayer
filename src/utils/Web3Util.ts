import {
	assetDataUtils,
	BigNumber,
	ContractWrappers,
	generatePseudoRandomSalt,
	orderHashUtils,
	RPCSubprovider,
	signatureUtils,
	SignedOrder,
	SignerType,
	Web3ProviderEngine
} from '0x.js';
import { schemas, SchemaValidator } from '@0xproject/json-schemas';
import { MnemonicWalletSubprovider } from '@0xproject/subproviders';
import { Web3Wrapper } from '@0xproject/web3-wrapper';
import * as CST from '../common/constants';
import { IRawOrder, IStringSignedOrder } from '../common/types';
import infura from '../keys/infura.json';
import util from './util';

export enum Wallet {
	None,
	Local,
	MetaMask,
	Ledger
}

export default class Web3Util {
	public contractWrappers: ContractWrappers;
	public web3Wrapper: Web3Wrapper;
	public wallet: Wallet = Wallet.None;
	public accountIndex: number = 0;

	constructor(window: any, live: boolean, mnemonic: string) {
		if (window && typeof window.web3 !== 'undefined') {
			this.web3Wrapper = new Web3Wrapper(window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else {
			const pe = new Web3ProviderEngine();
			pe.addProvider(
				new RPCSubprovider(
					(live ? CST.PROVIDER_INFURA_MAIN : CST.PROVIDER_INFURA_KOVAN) +
						'/' +
						infura.token
				)
			);
			if (!window && mnemonic) {
				const mnemonicWallet = new MnemonicWalletSubprovider({
					mnemonic: mnemonic,
					baseDerivationPath: CST.BASE_DERIVATION_PATH
				});
				pe.addProvider(mnemonicWallet);
			}
			pe.start();
			this.web3Wrapper = new Web3Wrapper(pe);
			this.wallet = window ? Wallet.None : Wallet.Local;
		}

		this.contractWrappers = new ContractWrappers(this.web3Wrapper.getProvider(), {
			networkId: live ? CST.NETWORK_ID_MAIN : CST.NETWORK_ID_KOVAN
		});
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		if (this.wallet !== Wallet.MetaMask) return;

		const store = (this.web3Wrapper.getProvider() as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public async getCurrentAddress(): Promise<string> {
		const accounts = await this.web3Wrapper.getAvailableAddressesAsync();
		return accounts[this.accountIndex] || CST.DUMMY_ADDR;
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3Wrapper.getNetworkIdAsync();
	}

	public async createRawOrder(
		userAddr: string,
		relayerAddr: string,
		makerAssetAddr: string,
		takerAssetAddr: string,
		makerAmt: number,
		takerAmt: number,
		expInSeconds: number
	): Promise<IRawOrder> {
		if (this.wallet !== Wallet.MetaMask) Promise.reject('cannot sign');
		const order = {
			senderAddress: relayerAddr.toLowerCase(),
			makerAddress: userAddr.toLowerCase(),
			takerAddress: relayerAddr.toLowerCase(),
			makerFee: new BigNumber(0),
			takerFee: new BigNumber(0),
			makerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(makerAmt), 18),
			takerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(takerAmt), 18),
			makerAssetData: assetDataUtils.encodeERC20AssetData(makerAssetAddr),
			takerAssetData: assetDataUtils.encodeERC20AssetData(takerAssetAddr),
			salt: generatePseudoRandomSalt(),
			exchangeAddress: this.contractWrappers.exchange.getContractAddress(),
			feeRecipientAddress: relayerAddr.toLowerCase(),
			expirationTimeSeconds: new BigNumber(expInSeconds)
		};

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const signature = await signatureUtils.ecSignOrderHashAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			order.makerAddress,
			SignerType.Metamask
		);
		return {
			orderHash: orderHash,
			signedOrder: { ...order, signature }
		};
	}

	public static getRandomFutureDateInSeconds = () => {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	};

	public static stringToBN = (value: string): BigNumber => {
		return new BigNumber(value);
	};

	public static fromWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toUnitAmount(new BigNumber(value), 18).valueOf());
	};

	public static toWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toWei(new BigNumber(value)).valueOf());
	};

	public static assetDataToTokenName = (assetData: string): string => {
		const tokenAddr = assetDataUtils.decodeERC20AssetData(assetData).tokenAddress;
		return CST.TOKEN_MAPPING[tokenAddr];
	};

	public static getSideFromSignedOrder = (
		order: SignedOrder | IStringSignedOrder,
		pair: string
	): string => {
		return Web3Util.assetDataToTokenName(order.takerAssetData) === pair.split('-')[0]
			? CST.DB_BID
			: CST.DB_ASK;
	};

	public static getPriceFromSignedOrder = (order: IStringSignedOrder, side: string): number => {
		const isBid = side === CST.DB_BID;
		return util.round(
			Web3Util.stringToBN(isBid ? order.makerAssetAmount : order.takerAssetAmount)
				.div(isBid ? order.takerAssetAmount : order.makerAssetAmount)
				.valueOf()
		);
	};

	public async validateOrder(signedOrder: SignedOrder): Promise<string> {
		const { orderSchema } = schemas;
		const { signature, ...order } = signedOrder;
		const validator = new SchemaValidator();
		if (!validator.validate(order, orderSchema).valid) {
			util.logDebug('invalid schema ' + JSON.stringify(signedOrder));
			return '';
		}

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const isValidSig = await signatureUtils.isValidSignatureAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			signature,
			order.makerAddress
		);
		if (!isValidSig) {
			util.logDebug('invalid signature ' + orderHash);
			return '';
		}

		return orderHash;
	}

	public getTokenAddressFromName(tokenName: string): string {
		switch (tokenName) {
			case CST.TOKEN_ZRX:
				return this.contractWrappers.exchange.getZRXTokenAddress();
			case CST.TOKEN_WETH:
				const ethTokenAddr = this.contractWrappers.etherToken.getContractAddressIfExists();
				if (!ethTokenAddr) {
					util.logInfo('no eth token address');
					return '';
				} else return ethTokenAddr;

			default:
				util.logInfo('no such token found');
				return '';
		}
	}

	public setAllUnlimitedAllowance(tokenAddr: string, addrs: string[]): Array<Promise<string>> {
		return addrs.map(address =>
			this.contractWrappers.erc20Token.setUnlimitedProxyAllowanceAsync(tokenAddr, address)
		);
	}

	public async setBaseQuoteAllowance(
		baseTokenAddr: string,
		quoteTokenAddr: string,
		addrs: string[]
	): Promise<void> {
		const responses = await Promise.all(
			this.setAllUnlimitedAllowance(quoteTokenAddr, addrs).concat(
				this.setAllUnlimitedAllowance(baseTokenAddr, addrs)
			)
		);
		await Promise.all(responses.map(tx => this.web3Wrapper.awaitTransactionSuccessAsync(tx)));
	}
}
