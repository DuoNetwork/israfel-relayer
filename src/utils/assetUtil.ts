import { assetDataUtils, BigNumber, SignedOrder } from '0x.js';
import { Web3Wrapper } from '@0xproject/web3-wrapper';
import { TransactionReceiptWithDecodedLogs } from 'ethereum-types';
import * as CST from '../common/constants';
import { IOption, IStringSignedOrder } from '../common/types';
import util from './util';
import Web3Util from './web3Util';
import { stringToBN } from './web3Util';

class AssetUtil {
	private web3Util: Web3Util | null = null;
	public makers: string[] = [];
	public taker: string = '';

	public async init(web3Util: Web3Util) {
		this.web3Util = web3Util;
		const [taker, ...makers] = await this.web3Util.web3Wrapper.getAvailableAddressesAsync();
		this.taker = taker;
		this.makers = makers;
	}

	public getRandomMaker(): string {
		const index = Math.floor(Math.random() * Math.floor(this.makers.length));
		return this.makers[index];
	}

	public async approveAllMakers(tokenAddress: string) {
		if (!this.web3Util) return;
		// Allow the 0x ERC20 Proxy to move erc20 token on behalf of makerAccount
		for (const maker of this.makers) {
			const makerZRXApprovalTxHash = await this.web3Util.contractWrappers.erc20Token.setUnlimitedProxyAllowanceAsync(
				tokenAddress,
				maker
			);
			await this.web3Util.web3Wrapper.awaitTransactionSuccessAsync(makerZRXApprovalTxHash);
		}
	}

	public assetDataToTokenName(assetData: string): string {
		const tokenAddr = assetDataUtils.decodeERC20AssetData(assetData).tokenAddress;
		return CST.TOKEN_MAPPING[tokenAddr];
	}

	public getSideFromSignedOrder(order: SignedOrder | IStringSignedOrder, pair: string): string {
		return this.assetDataToTokenName(order.takerAssetData) === pair.split('-')[0]
			? CST.DB_BID
			: CST.DB_ASK;
	}

	public getPriceFromSignedOrder(order: IStringSignedOrder, side: string): number {
		if (![CST.DB_BID, CST.DB_ASK].includes(side)) throw new Error('wrong side specified');
		return side === CST.DB_BID
			? util.round(
					stringToBN(order.takerAssetAmount)
						.div(order.makerAssetAmount)
						.valueOf()
			)
			: util.round(
					stringToBN(order.makerAssetAmount)
						.div(order.takerAssetAmount)
						.valueOf()
			);
	}

	public getTokenAddressFromName(tokenName: string): string {
		if (!this.web3Util) return '';
		switch (tokenName) {
			case CST.TOKEN_ZRX:
				return this.web3Util.contractWrappers.exchange.getZRXTokenAddress();
			case CST.TOKEN_WETH:
				const ethTokenAddr = this.web3Util.contractWrappers.etherToken.getContractAddressIfExists();
				if (!ethTokenAddr) {
					util.logInfo('no eth token address');
					return '';
				} else return ethTokenAddr;

			default:
				util.logInfo('no such token found');
				return '';
		}
	}

	public async setTokenAllowance(option: IOption): Promise<TransactionReceiptWithDecodedLogs> {
		if (!this.web3Util) throw new Error('error');
		const TxHash = await this.web3Util.contractWrappers.erc20Token.setAllowanceAsync(
			this.getTokenAddressFromName(option.token),
			this.makers[option.maker],
			option.spender
				? this.makers[option.spender]
				: this.web3Util.contractWrappers.exchange.getContractAddress(),
			Web3Wrapper.toBaseUnitAmount(new BigNumber(option.amount), 18)
		);
		return await this.web3Util.web3Wrapper.awaitTransactionSuccessAsync(TxHash);
	}

	public setAllUnlimitedAllowance(tokenAddr: string, addrs: string[]): Array<Promise<string>> {
		return addrs.map(
			address =>
				this.web3Util
					? this.web3Util.contractWrappers.erc20Token.setUnlimitedProxyAllowanceAsync(
							tokenAddr,
							address
					)
					: Promise.reject()
		);
	}

	public async setBaseQuoteAllowance(
		baseTokenAddr: string,
		quoteTokenAddr: string,
		addrs: string[]
	): Promise<void> {
		const responses = await Promise.all(
			this.setAllUnlimitedAllowance(quoteTokenAddr, addrs).concat(
				this.setAllUnlimitedAllowance(baseTokenAddr, addrs)
			)
		);
		await Promise.all(
			responses.map(
				tx =>
					this.web3Util
						? this.web3Util.web3Wrapper.awaitTransactionSuccessAsync(tx)
						: Promise.reject()
			)
		);
	}

	// TODO add from signedOrder to orderHash function
}
const assetUtil = new AssetUtil();
export default assetUtil;
