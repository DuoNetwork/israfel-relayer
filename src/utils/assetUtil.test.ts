// import {BigNumber} from '0x.js';
// import { Web3Wrapper } from '@0xproject/web3-wrapper';
import { IStringSignedOrder } from '../common/types';
import assetUtil from './assetUtil';
import Web3Util from './web3Util';

const web3Util = new Web3Util();

test('getTokenAddressFromName', async () => {
	await assetUtil.init(web3Util);
	expect(assetUtil.getTokenAddressFromName('ZRX')).toMatchSnapshot();
	expect(assetUtil.getTokenAddressFromName('WETH')).toMatchSnapshot();
	expect(assetUtil.getTokenAddressFromName('')).toMatchSnapshot();
});

const rawOrder: IStringSignedOrder = {
	exchangeAddress: '0x48bacb9266a570d521063ef5dd96e61686dbe788',
	makerAddress: '0xa8dda8d7f5310e4a9e24f8eba77e091ac264f872',
	takerAddress: '0x5409ed021d9299bf6814279a6a1411a7e866a631',
	senderAddress: '0xa8dda8d7f5310e4a9e24f8eba77e091ac264f872',
	feeRecipientAddress: '0x5409ed021d9299bf6814279a6a1411a7e866a631',
	expirationTimeSeconds: 'BigNumber { s": 1, e": 9, c": [ 1538117918 ] }',
	salt:
		'BigNumber {s": 1,e": 76,c":[ 4819806,21385749514209,88700844036866,26535141779564,49454803591044,15105341483720 ] }',
	makerAssetAmount: 'BigNumber { s": 1, e": 18, c": [ 63530 ] }',
	takerAssetAmount: 'BigNumber { s": 1, e": 17, c": [ 8450 ] }',
	makerAssetData: '0xf47261b0000000000000000000000000871dd7c2b4b25e1aa18728e9d5f2af4c4e431f5c',
	takerAssetData: '0xf47261b00000000000000000000000000b1ba0af832d7c05fd64161e0db78e85978e8082',
	makerFee: 'BigNumber { s": 1, e": 0, c": [ 0 ] }',
	takerFee: 'BigNumber { s": 1, e": 0, c": [ 0 ] }',
	signature: ''
};
const pair = 'ZRX-WETH';

// test('getSideFromSignedOrder', async () => {
// 	expect(assetUtil.getSideFromSignedOrder(rawOrder, pair)).toMatchInlineSnapshot();
// });
