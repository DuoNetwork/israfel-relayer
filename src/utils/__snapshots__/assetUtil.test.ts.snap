// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenAddressFromName 1`] = `"0x2002d3812f58e35f0ea1ffbf80a75a38c32175fa"`;

exports[`getTokenAddressFromName 2`] = `"0xd0a1e359811322d97991e03f863a0c30c2cf029c"`;

exports[`getTokenAddressFromName 3`] = `""`;
